From 7482a34f9e5eac9e1bf9d92a96e7698adc4ed026 Mon Sep 17 00:00:00 2001
From: Michal Schmidt <mschmidt@redhat.com>
Date: Fri, 5 Sep 2014 19:49:45 -0400
Subject: [ethernet] bnx2x: Enlarge the dorq threshold for VFs

Message-id: <1409946615-5862-41-git-send-email-mschmidt@redhat.com>
Patchwork-id: 91580
O-Subject: [RHEL7.1 PATCH 40/70] bnx2x: Enlarge the dorq threshold for VFs
Bugzilla: 1089402
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1089402

Upstream commit:
commit f2cfa997ef77bcfc3900eb968fa5254f951bb8f2
Author: Ariel Elior <ariel.elior@qlogic.com>
Date:   Thu Jun 12 07:55:32 2014 +0300

    bnx2x: Enlarge the dorq threshold for VFs

    A malicious VF might try to starve the other VFs & PF by creating
    contineous doorbell floods. In order to negate this, HW has a threshold of
    doorbells per client, which will stop the client doorbells from arriving
    if crossed.

    The threshold currently configured for VFs is too low - under extreme traffic
    scenarios, it's possible for a VF to reach the threshold and thus for its
    fastpath to stop working.

    Signed-off-by: Ariel Elior <ariel.elior@qlogic.com>
    Signed-off-by: Yuval Mintz <yuval.mintz@qlogic.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_sriov.c b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_sriov.c
index 4d32389..adecc1d 100644
--- a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_sriov.c
+++ b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_sriov.c
@@ -1071,8 +1071,10 @@ void bnx2x_iov_init_dq(struct bnx2x *bp)
  REG_WR(bp, DORQ_REG_VF_TYPE_MIN_MCID_0, 0);
  REG_WR(bp, DORQ_REG_VF_TYPE_MAX_MCID_0, 0x1ffff);
 
- /* set the VF doorbell threshold */
- REG_WR(bp, DORQ_REG_VF_USAGE_CT_LIMIT, 4);
+ /* set the VF doorbell threshold. This threshold represents the amount
+  * of doorbells allowed in the main DORQ fifo for a specific VF.
+  */
+ REG_WR(bp, DORQ_REG_VF_USAGE_CT_LIMIT, 64);
 }
 
 void bnx2x_iov_init_dmae(struct bnx2x *bp)
-- 
1.7.1